subseas_flex_fcst/maproom.py

```diff
@@ -131,40 +131,63 @@ def write_mapTitle(startDate, leadTime, leadTimeOptions):
     Output("lngInput", "value"),
     Input("submitLatLng","n_clicks"),
     Input("map", "click_lat_lng"),
-    Input("startDate","value"),
-    Input("leadTime","value"),
     State("latInput", "value"),
     State("lngInput", "value")
 )
-def local_plots(n_clicks, click_lat_lng, startDate, leadTime, latitude, longitude):
+def pick_location(n_clicks, click_lat_lng, latitude, longitude):
     # Reading
-
-    fcst_mu, fcst_var, obs, hcst = read_cptdataset(leadTime, startDate, y_transform=CONFIG["y_transform"])
-    if click_lat_lng is None: #Map was not clicked
-        if n_clicks == 0: #Button was not clicked (that's landing page)
-            lat = (fcst_mu.Y[0].values+fcst_mu.Y[-1].values)/2
-            lng = (fcst_mu.X[0].values+fcst_mu.X[-1].values)/2
-        else: #Button was clicked
+    filesNameList = glob.glob(f'{DATA_PATH}/{CONFIG["forecast_mu_filePattern"]}')
+    startDate = re.search("\w{3}-\w{1,2}-\w{4}",filesNameList[0])
+    startDate = datetime.strptime(startDate.group(),"%b-%d-%Y").strftime("%b-%-d-%Y")
+    fcst_mu = predictions.selFile(
+        DATA_PATH,
+        CONFIG["forecast_mu_filePattern"],
+        list(CONFIG["leads"])[0],
+        startDate
+    )
+    if dash.ctx.triggered_id == None:
+        lat = fcst_mu.Y[int(fcst_mu.Y.size/2)].values
+        lng = fcst_mu.X[int(fcst_mu.X.size/2)].values
+    else:
+        if dash.ctx.triggered_id == "map":
+            lat = click_lat_lng[0]
+            lng = click_lat_lng[1]
+        else:
             lat = latitude
             lng = longitude
-    else: #Map was clicked
-        lat = click_lat_lng[0]
-        lng = click_lat_lng[1]
+        half_res = (fcst_mu["X"][1] - fcst_mu["X"][0]) / 2
+        tol = np.sqrt(2 * np.square(half_res)).values
+        try:
+            nearest_grid = fcst_mu.sel(X=lng, Y=lat, method="nearest", tolerance=tol)
+            lat = nearest_grid.Y.values
+            lng = nearest_grid.X.values
+        except KeyError:
+            lat = lat
+            lng = lng
+    return dlf.Marker(position=[lat, lng]), lat, lng
+
+
+@APP.callback(
+    Output("cdf_graph", "figure"),
+    Output("pdf_graph", "figure"),
+    Input("layers_group", "children"),
+    Input("startDate","value"),
+    Input("leadTime","value"),
+)
+def local_plots(marker, startDate, leadTime):
+    # Reading
+    lat = marker["props"]["position"][0]
+    lng = marker["props"]["position"][1]
+    fcst_mu, fcst_var, obs, hcst = read_cptdataset(leadTime, startDate, y_transform=CONFIG["y_transform"])
     # Errors handling
     try:
         half_res = (fcst_mu["X"][1] - fcst_mu["X"][0]) / 2
         tol = np.sqrt(2 * np.square(half_res)).values
-        nearest_grid = fcst_mu.sel(X=lng, Y=lat, method="nearest", tolerance=tol)
-        lat = nearest_grid.Y.values
-        lng = nearest_grid.X.values
         fcst_mu = fcst_mu.sel(X=lng, Y=lat, method="nearest", tolerance=tol)
         fcst_var = fcst_var.sel(X=lng, Y=lat, method="nearest", tolerance=tol)
```

Comment:
We already snapped the selection to the grid when setting the marker position, right? So can't we just do `.sel` without a tolerance here?

If we do need to re-snap it here, then factor out the logic that this function and pick_location have in common.

Reply:
I don't snap when the click/input is outside the data domain. When outside the data domain, pick_location returns whatever lon/lat was clicked/input   and local_plots triggers the errors. So either we rethink of the behavior of outside domain click/input (I don't have an idea right now; either I factor out, as you said.

Reply:
Interesting--the same `.sel` operation does double duty: in one place it snaps the location to the grid, and in the other place it checks whether the location is in or out of bounds. 

We could change the behavior of clicking out of bounds--instead of moving the marker and displaying the error charts, we could just refuse to move the marker there. Not sure, that might be confusing/look like a bug.

Probably just factor it out.

Reply:
We could that. That is if click/input out of bounds: do nothing. We could also snap to the closest location still, that would be on the edge of the domain. It might look funny to a user... on the other hand there won't be out of bounds error graphs anymore.

In the future, we might have apps with layers of different resolution (e.g. grids, geometries, points), where we would have more complex behavior, e.g. snap to grid and get closest point to overlay time series from both (that might be the case already in DBflood...). So maybe it's safer to keep the current behavior... thinking out loud now.

Reply:
I've just pushed 2 commits, the second one is trying to do that. Not sure I got what you had in mind so please have a look.